builder/pipeline.cc

```diff
@@ -258,118 +263,164 @@ bool operator==(const loop_id& a, const loop_id& b) {
 }
 
 class pipeline_builder {
-  // We're going to incrementally build the body, starting at the end of the pipeline and adding
-  // producers as necessary.
-  std::set<buffer_expr_ptr> to_produce;
-  std::list<buffer_expr_ptr> to_allocate;
-  std::set<buffer_expr_ptr> produced, consumed;
   std::set<buffer_expr_ptr> allocated;
 
-  stmt result;
-
-public:
-  pipeline_builder(const std::vector<buffer_expr_ptr>& inputs, const std::vector<buffer_expr_ptr>& outputs,
-      std::set<buffer_expr_ptr>& constants) {
-    // To start with, we need to produce the outputs.
-    for (auto& i : outputs) {
-      to_produce.insert(i);
-      allocated.insert(i);
-    }
-    for (auto& i : inputs) {
-      produced.insert(i);
-    }
-
-    // Find all the buffers we need to produce.
-    while (true) {
-      std::set<buffer_expr_ptr> produce_next;
-      for (const buffer_expr_ptr& i : to_produce) {
-        if (!i->producer()) {
-          // Must be an input.
-          continue;
-        }
-
-        for (const func::input& j : i->producer()->inputs()) {
-          if (!to_produce.count(j.buffer)) {
-            if (j.buffer->constant()) {
-              constants.insert(j.buffer);
-            } else {
-              produce_next.insert(j.buffer);
-            }
-          }
-        }
-      }
-      if (produce_next.empty()) break;
-
-      to_produce.insert(produce_next.begin(), produce_next.end());
-    }
-  }
-
-  // f can be called if it doesn't have an output that is consumed by a not yet produced buffer's producer.
-  bool can_produce(const func* f) const {
-    for (const buffer_expr_ptr& p : to_produce) {
-      if (produced.count(p)) {
-        // This buffer is already produced.
+  // Topologically sorted functions.
+  std::vector<const func*> order_;
+  // A mapping between func's and their compute_at locations.
+  std::map<const func*, loop_id> compute_at_levels_;
+
+  void topological_sort_impl(const func* f, std::set<const func *>& visited,
```

Comment:
Should this be static, or a free function? I don't think it accesses any members. Same question goes for several of the helpers below

Reply:
I had them as a free functions, but for some unknown reason put them into the class in the end. Changed back.